src/jinja2/utils.py

```diff
@@ -237,71 +231,85 @@ def urlize(text, trim_url_limit=None, rel=None, target=None, extra_schemes=None)
         or without the ``mailto:`` scheme. Validate IP addresses. Ignore
         parentheses and brackets in more cases.
     """
+    if trim_url_limit is not None:
+
+        def trim_url(x):
+            if len(x) > trim_url_limit:
+                return f"{x[:trim_url_limit]}..."
+
+            return x
 
-    def trim_url(x, limit=trim_url_limit):
-        if limit is not None:
-            return x[:limit] + ("..." if len(x) >= limit else "")
+    else:
 
-        return x
+        def trim_url(x):
+            return x
 
-    words = _word_split_re.split(str(escape(text)))
+    words = re.split(r"(\s+)", str(escape(text)))
     rel_attr = f' rel="{escape(rel)}"' if rel else ""
     target_attr = f' target="{escape(target)}"' if target else ""
 
     for i, word in enumerate(words):
-        match = _punctuation_re.match(word)
+        head, middle, tail = "", word, ""
+        match = re.match(r"^([(<]|&lt;)+", middle)
```

Comment:
This calls `re.match()` with the same pattern string for each iteration of the loop (and similar is done below). I don’t think the `re` module caches recently- or frequently-compiled patterns, in which case these calls would be needlessly recompiling the same pattern in every iteration, is that right? If so, is it worth compiling these patterns once, and reusing the resulting pattern objects here?

Reply:
`re` caches the last 512 patterns used. Compiling every single pattern at import time increases startup time for projects, I'd seen some more (general) discussions about this recently so decided these could be inlined.

Reply:
Oh interesting, good to know, thanks!